[RFC] semver (semantic versioning) compliance

justinmk opened the pull request:
[removed outdated comments, see comments below]

justinmk committed:
version: semver.org compliance

We use `git describe` to stamp pre-release versions (dev builds). But
`git describe` uses the result of the most-recent tag (the current
_release_ version)--so we must munge it with the _next_ (i.e.
unreleased) version.

Also fix non-git builds: do not invoke git_describe(NVIM_VERSION_MEDIUM)
if we're not in a git tree, else it gets the dummy value
"HEAD-HASH-NOTFOUND".

Example :version output in non-git build:
  NVIM 0.1.2-dev

Example :version output in git build:
  NVIM v0.1.2-176-g9c3c2b5

justinmk commented on CMakeLists.txt:
```diff
@@ -51,6 +51,14 @@ endif()
+#   cmake -DNVIM_VERSION_PRERELEASE=
+if(NOT DEFINED NVIM_VERSION_PRERELEASE)
+  set(NVIM_VERSION_PRERELEASE "-dev")
+endif()
```
@jamessan @fwalch Note regarding packaging: if you build in a git tree, then this may not affect you. If you don't, then the `-dev` suffix will be appended to the version string unless you define `NVIM_VERSION_PRERELEASE` to be empty.

This avoids having to physically remove `NVIM_VERSION_PRERELEASE` in our source (e.g. 4a1c36e953559382362b79be7630a64163c43ef2) every time we tag a release.

Pyrohh replied:
The 2nd sentence would be helpful in the commit message.

fwalch replied:
I don't like this too much.. I think the version should not depend on how Nvim is built. If a source snapshot is not a release, it should have "-dev", otherwise it shouldn't. I agree that it's cumbersome to remove `NVIM_VERSION_PRERELEASE` when we tag a release, though.. but I don't have a better idea.

If a packager modifies code or runtime files via patches, they can as well do that with the CMake code, so we don't need to expose a CMake property for this use-case.

Semi-related: currently we number versions like: `0.1.0 -> 0.1.0-dev -> 0.1.1 -> 0.1.1-dev -> ...` This doesn't follow SemVer:

> When major, minor, and patch are equal, a pre-release version has lower precedence than a normal version. Example: 1.0.0-alpha < 1.0.0.

If we number our versions like `0.1.0 -> 0.1.1-dev -> 0.1.1 -> 0.1.2-dev -> ...`, we anyway have to add a follow-up commit after tagging a release, to increase the version number. This commit could also re-introduce a hardcoded `NVIM_VERSION_PRERELEASE`.

I didn't know how other projects do this, so I picked `libuv` more or less at random.. looks like they do what I described: [tagging a release](https://github.com/libuv/libuv/commit/5467299450ecf61635657557b6e01aaaf6c3fdf4#diff-6da6ac057a6e89d11bde410b064ec16eL34), [follow-up commit](https://github.com/libuv/libuv/commit/61f13a6d5e52817c748ddedea73efcf3eb94c089)

Now that I think of it, what I described so far only applies to "building from tarballs"... `git describe` complicates the situation somewhat. We don't really follow SemVer anymore since we introduced that. Hm.

justinmk replied:
> If a source snapshot is not a release, it should have "-dev", otherwise it shouldn't.

AFAIK that is how this PR functions. In which case would that not be the case (except in a git tree)? 

> Semi-related: currently we number versions like: 0.1.0 -> 0.1.0-dev -> 0.1.1 -> 0.1.1-dev -> ... This doesn't follow SemVer:

Ok, what if we did `0.1.0 -> 0.1.1-dev -> 0.1.1 -> 0.1.2-dev -> ...` ?

> libuv more or less at random.. looks like they do what I described: tagging a release, follow-up commit

I tend to look at git, which AFAICT does not do that. We've been burned before by following node.js habits (e.g.: CLA). How about other projects (e.g., Rust, Atom--I did not check these myself)?

> git describe complicates the situation somewhat. We don't really follow SemVer anymore since we introduced that. Hm.

Section 9 of semver.org says arbitrary pre-release data may be appended after a hyphen...

fwalch replied:
> AFAIK that is how this PR functions. In which case would that not be the case (except in a git tree)? 

You can build any revision with `-DNVIM_VERSION_PRERELEASE=`, which would make it look like a release. We hardcode the version number, so I feel like we should also hardcode this part of the version and make it do "the right thing" out of the box. It's more of a personal preference thing, though.

> How about other projects (e.g., Rust, Atom--I did not check these myself)?

I will check some more projects.

> Section 9 of semver.org says arbitrary pre-release data may be appended after a hyphen...

Yes, but git describe appends "post-release data", i.e. it's supposed to be 0.1.0 < 0.1.0-1, whereas semver would treat it as 0.1.0-1 < 0.1.0, as the pre-release data after the hyphen is a string, not a number (see also https://github.com/mojombo/semver/issues/106). Also, if we do the 0.1.0 -> 0.1.1-dev versioning scheme, then we would have a version of e.g. 0.1.1-dev for tarball builds and 0.1.0-5-abcdef for `git describe` builds.. maybe we should not use git describe directly, but e.g. `0.1.1-dev` for tarball builds and `0.1.1-dev+abcdef` for Git tree builds? :grimacing:

If we decide to change anything, I would create the PR, so you don't have to do all the work.. I hope I'm not overcomplicating things :grimacing:

jszakmeister replied:
> I don't like this too much.. I think the version should not depend on how Nvim is built. If a source snapshot is not a release, it should have "-dev", otherwise it shouldn't. I agree that it's cumbersome to remove NVIM_VERSION_PRERELEASE when we tag a release, though.. but I don't have a better idea.

Another idea--that might be slightly better--is to include an extra file in the release, say .version-info.  We can then have CMake pickup on this file and coerce the version number (leaving off the "-dev", for instance), and when it's not there keep the current behavior.

We use a similar mechanism at our workplace so that we can control the version number via our QuickBuild instance, and it allows us to promote the builds through to a final release--ultimately dropping things out of the version that mark it as a dev or a release candidate.  We also use it to label our dev versions separately than the git commit numbers--because sometimes there are packaging issues and you need to rebuild from the same commit number, but it's not the same build.

The downside is that there's an extra file you need to include as part of the tag, but at least you don't have to remember to pass yet another argument.

Pyrohh commented:
Besides the nit LGTM.

Pyrohh commented:
No comment on `NVIM_VERSION_PRERELEASE`, but the `:version` documentation should also be updated:
```
                                                        *:ve* *:version*
:ve[rsion]              Print the version number of the editor.  If the
                        compiler used understands "__DATE__" the compilation
                        date is mentioned.  Otherwise a fixed release-date is
                        shown.
                        ...
```

justinmk committed:
BuildLibuv.cmake: make case consistent

justinmk commented:
Updated.

> git describe complicates the situation somewhat. We don't really follow SemVer anymore since we introduced that. Hm.

This is fixed now: the version part of `git describe` is replaced with the hardcoded `NVIM_VERSION_*` values.

> then we would have a version of e.g. 0.1.1-dev for tarball builds and 0.1.0-5-abcdef for git describe builds.. 

With the latest change it will be 0.1.1-dev for tarball builds and 0.1.1-5-abcdef for git builds. That seems reasonable to me.

> maybe we should not use git describe directly, but e.g. 0.1.1-dev for tarball builds and 0.1.1-dev+abcdef for Git tree builds?

Per semver.org, "Build metadata SHOULD be ignored when determining version precedence", so it can't be used to describe different sources.

justinmk commented on CMakeLists.txt:
```diff
@@ -51,15 +51,25 @@ if(NOT CMAKE_BUILD_TYPE)
 include(GetGitRevisionDescription)
-git_describe(NVIM_VERSION_MEDIUM)
+get_git_head_revision(GIT_REFSPEC NVIM_VERSION_COMMIT)
+if(NVIM_VERSION_COMMIT) # is a git repo
```
@Pyrohh I had to restore this condition because otherwise NVIM_VERSION_MEDIUM gets a dummy value "HEAD-HASH-NOTFOUND".

re 7be12edd38b1da0a506f39651d8c6a6f9aa5626a

Pyrohh replied:
Ah, sounds good to me. Thanks for updating the Following-Head page, too.

fwalch commented:
> Per semver.org, "Build metadata SHOULD be ignored when determining version precedence", so it can't be used to describe different sources.

Maybe I misunderstand what you're saying, but isn't that what we want (same versions for different sources)? Now, we have different versions: according to semver, `0.1.1-dev != 0.1.1-5-abcdef`. By specifying the commit as build metadata, we would have `0.1.1-dev == 0.1.1-dev+abcdef`.

justinmk commented:
> By specifying the commit as build metadata, we would have `0.1.1-dev == 0.1.1-dev+abcdef`.

But builds based on different source code _are not_ the same, so wouldn't it be misleading to imply that 

    0.1.1-dev == 0.1.1-dev+abcdef == 0.1.1-dev+ghijklm == 0.1.1-dev+xyzabcd == ...

? If you think it's acceptable then that's fine with me. Though I would like to use the `git describe` output so that it looks like this:

    0.1.1+5-gabcdef.

Otherwise we are stuck with the "pre release" workaround. Also, the "N commits since" number is useful to see at a glance how from stable one is.

fwalch commented:
> But builds based on different source code are not the same, so wouldn't it be misleading to imply

You're right.. that's also the situation we have with the tarballs now, but at least for Git-based builds we can resolve it.

So I guess both approaches have pros and cons.. I'm not sure which one is better, so let's leave it as is.
